File: ios/Sources/CertificateInfoPlugin/CertificateInfoPlugin.swift
import Foundation
import Capacitor

@objc(CertificateInfoPlugin)
public class CertificateInfoPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "CertificateInfoPlugin"
    public let jsName = "CertificateInfo"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "getExpirationDate", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getExpirationDates", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "checkValidation", returnType: CAPPluginReturnPromise),
    ]
  private var implementation: CertificateInfo?

  override public func load() {
    self.implementation = CertificateInfo(config: getCertificateInfoConfig())
  }
  
  @objc func getExpirationDate(_ call: CAPPluginCall) {
          do {
            let filename: String = (call.getString("filename") ?? "").trimmingCharacters(in: .whitespacesAndNewlines);
            
            let certData = try implementation!.getExpirationDate(from: filename);
            
            call.resolve([
              "expirationDate": implementation!.formatDateToISO(certData),
            ])
          } catch let error as CertificateInforError {
            let errorMessage = CertificateInfoErrorMessages.getErrorMessage(error: error)
            call.reject(errorMessage, "500")
            
          }catch {
            call.reject("Uknown error", "500")
        }
      }
  
  @objc func getExpirationDates(_ call: CAPPluginCall) {
          do {
            var expirationStringDates: [String] = []
            let expirationDates = try implementation!.getExpirationDates()
            
            for expirationDate in expirationDates {
              expirationStringDates.append(implementation!.formatDateToISO(expirationDate))
            }
            
            call.resolve([
              "expirationDates": expirationStringDates,
            ])
          } catch let error as CertificateInforError {
            let errorMessage = CertificateInfoErrorMessages.getErrorMessage(error: error)
            call.reject(errorMessage, "500")
            
          }catch {
            call.reject("Uknown error", "500")
        }
      }
  
  @objc func checkValidation(_ call: CAPPluginCall){
    do {
      let dateToCompare = call.getString("date");
      
      let certData = try implementation!.checkValidation(date: dateToCompare)
      
      call.resolve([
        "valid": certData,
      ])
    } catch let error as CertificateInforError {
      let errorMessage = CertificateInfoErrorMessages.getErrorMessage(error: error)
      call.reject(errorMessage, "500")
      
    }catch {
      call.reject("Uknown error", "500")
  }
  }
  
  
  private func getCertificateInfoConfig() -> CertificateInfoConfig {
    var config = CertificateInfoConfig();
    
    if let path = getConfig().getString("certificatesPath") {
      config.certificatePath = path
    }
    
    return config
  }
  
}

File: android/src/main/java/com/axity/plugin/certificate/CertificateInfoPlugin.java
package com.axity.plugin.certificate;

import android.util.Log;
import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import java.lang.reflect.Executable;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;
import org.json.JSONArray;
import org.json.JSONObject;

@CapacitorPlugin(name = "CertificateInfo")
public class CertificateInfoPlugin extends Plugin {

    private CertificateInfo implementation;

    @Override
    public void load() {
        CertificateInfoConfig config = getCertificateInfoConfig();

        implementation = new CertificateInfo(getContext(), config);
    }

    @PluginMethod
    public void getExpirationDate(PluginCall call) {
        try {
            String filename = call.getString("filename");
            Date expirationDate = implementation.getExpirationDate(filename);

            JSObject response = new JSObject();
            response.put("expirationDate", implementation.formatDateToISO(expirationDate));

            call.resolve(response);
        } catch (CertificateInfoException e) {
            call.reject(e.getMessage(), "500");
        } catch (Exception e) {
            call.reject("Uknown error", "500");
        }
    }

    @PluginMethod
    public void getExpirationDates(PluginCall call) {
        try {
            List<String> expirationStringDates = new ArrayList<>();
            List<Date> expirationDates = implementation.getExpirationDates();

            for (Date exp : expirationDates) {
                expirationStringDates.add(implementation.formatDateToISO(exp));
            }

            JSObject response = new JSObject();
            response.put("expirationDates", new JSONArray(expirationStringDates));

            call.resolve(response);
        } catch (CertificateInfoException e) {
            call.reject(e.getMessage(), "500");
        } catch (Exception e) {
            call.reject("Uknown error", "500");
        }
    }

    @PluginMethod
    public void checkValidation(PluginCall call) {
        try {
            String date = call.getString("date", "");
            boolean valid = implementation.checkValidation(date);
            JSObject response = new JSObject();
            response.put("valid", valid);

            call.resolve(response);
        } catch (CertificateInfoException e) {
            call.reject(e.getMessage(), "500");
        } catch (Exception e) {
            call.reject("Uknown error", "500");
        }
    }

    private CertificateInfoConfig getCertificateInfoConfig() {
        CertificateInfoConfig config = new CertificateInfoConfig();

        String certificatePath = getConfig().getString("certificatesPath", config.getCertificatePath());
        config.setCertificatePath(certificatePath);

        String dateFormat = getConfig().getString("dateFormat", config.getDateFormat());
        config.setDateFormat(dateFormat);

        return config;
    }
}
